Add `fn_kwargs` to `map` and `filter` of `IterableDataset` and `IterableDatasetDict`

yuukicammy opened the pull request:
# Overview
I've added an argument`fn_kwargs` for map and filter methods of `IterableDataset` and `IterableDatasetDict` classes.

# Details
Currently, the map and filter methods of some classes related to `IterableDataset` do not allow specifing the arguments passed to the function. This pull request adds `fn_kwargs` to pass arguments to the mapping function. This allows users to preprocess data more flexibly.

Added `fn_kwargs` to the following classes and methods (description of the argument is also added).
1. class `FilteredExamplesIterable`
2. method `filter` of class `IterableDataset`
3. method `map` of class `IterableDatasetDict`
4. method `filter` of class `IterableDatasetDict`

# Example of changes
Here's an example of how to use the new functionality:
```python
from datasets import IterableDatasetDict

def preprocess_function(example, a=None, b=None):
    # do something
    return example

dataset = IterableDatasetDict(...)
dataset = dataset.map(preprocess_function, fn_kwargs={"a": 1, "b": 2})
```
# Related Issues
This pull request is related to the following issue:
https://github.com/huggingface/datasets/issues/3444 .

# Testing
I have added unit tests to test the new functionality.

In test_iterable_dataset.py  
- Added `test_filtered_examples_iterable_with_fn_kwargs` for [1](#details).
- Added `test_iterable_dataset_filter` for [2](#details).
- Added `test_iterable_dataset_map_with_fn_kwargs`. This is not a newly added feature, but was added because it was not tested.

In test_dataset_dict.py  
- Added `_create_dummy_iterable_dataset` for [3](#details) and [4](#details).
- Added `_create_dummy_iterable_dataset_dict` for [3](#details) and [4](#details).
- Added `test_iterable_map` for [3](#details).
- Added `test_iterable_filter` for [4](#details).

Note that, there is no test for `IterableDatasetDict` at the current main branch. I thought about writing tests for `IterableDatasetDict` in a new file, but I decided to add them in the test file for `DatasetDict` (test_dataset_dict.py).

# Checklist
- [x] Format the code.
- [x] Added tests.
- [x] Passed tests locally.

lhoestq commented on tests/test_iterable_dataset.py:
```diff
@@ -502,6 +502,42 @@ def test_filtered_examples_iterable_with_indices(n, func, batched, batch_size):
+        (5, lambda x, y: [i == 0 for i in x["id"]], True, -1, {"y": 1}),  # same with bs<=0
+    ],
+)
+def test_filtered_examples_iterable_with_fn_kwargs(n, func, batched, batch_size, fn_kwargs):
```
This test seems redundant with the previous ones, I think you can remove it and we can keep the other tests you added

yuukicammy replied:
I have addressed your comments. Please confirm that it is as intended.
9bd41d4dc2c1b544fa3efa8c7b1308340dbc5835

lhoestq replied:
Sorry I meant you can remove `test_filtered_examples_iterable_with_fn_kwargs` completely

lhoestq commented on tests/test_iterable_dataset.py:
```diff
@@ -841,6 +877,27 @@ def test_iterable_dataset_map_with_features(dataset: IterableDataset) -> None:
 
+def test_iterable_dataset_map_with_fn_kwargs(dataset: IterableDataset) -> None:
+    fn_kwargs = {"y": 1}
+    with dataset.map(lambda x, y: {"id+y": x["id"] + y}, fn_kwargs=fn_kwargs) as mapped_dataset:
```
No need to use `with` statements on `IterableDataset` since it doesn't use temporary files when `map` is called. You can remove the `with` statements in this test and the next one.

yuukicammy replied:
Thanks! I have modified them. 9bed249f8f33daff13c24b3d8a3c7cc95124eb1f

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/datasets/pr_5810). All of your documentation changes will be reflected on that endpoint.

yuukicammy commented:
Sorry, the local test passed because it was inadvertently testing the main branch. I am currently fixing where the test failed.

yuukicammy commented:
- I have fixed the bug and addressed the above two points.
- I have tested locally and confirmed that the test passes.

Please check the contents. @lhoestq 

5715a7e64bdd2951e6705aee58d592392e1538d6

lhoestq commented:
Cool ! You can run `make style` to fix code formatting to fix the ci

yuukicammy commented:
I had forgotten about it. I did it.
00248926a37c6f1387614aa388c36fdc105a59f5

brianhill11 commented:
Thanks for putting this together @yuukicammy ! Looking forward to using this new addition ASAP. 
@lhoestq - sorry to bother you with this, but if this looks good to you, any chance we could get this merged in? 

Thanks again to you both!

lhoestq commented:
Yup there's just one test to remove and we can merge

yuukicammy commented:
Sorry for my understanding wrong! Correspondence has been addressed.  @lhoestq 
 ca511b7b29fdde51ffd69b58bda79220472e9e94

Thanks for your comment! @brianhill11

lhoestq reviewed (approved):
Thanks :) LGTM
